src/libinjector/linux/linux_utils.c

```diff
@@ -0,0 +1,263 @@
+
+        if (j == M) // if pat[0...M-1] = txt[i, i+1, ...i+M-1]
+            return i;
+    }
+    return -1;
+}
+
+addr_t find_syscall(drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t vdso)
+{
+    // lock vmi
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_PID,
+        .pid = info->proc_data.pid,
+        .addr = vdso
+    );
+
+    size_t size = 4096;
+    size_t bytes_read = 0;
+    char* vdso_memory = g_try_malloc(size);
+
+    // read the vdso memory
+    bool success = (VMI_SUCCESS == vmi_read(vmi, &ctx, size, (void*)vdso_memory, &bytes_read));
+    if (!success)
+        fprintf(stderr, "Could not vdso memory\n");
+    else
+    {
+        PRINT_DEBUG("vdso memory read successful\n");
+    }
+
+    // release vmi
+    drakvuf_release_vmi(drakvuf);
+
+    char syscall[] = { 0xf, 0x5 };
+    size_t syscall_offset = search(vdso_memory, syscall, size, 2);
+    PRINT_DEBUG("syscall offset: %ld\n", syscall_offset);
+    PRINT_DEBUG("syscall addr: %lx\n", vdso + syscall_offset);
+
+    free(vdso_memory);
+
+    injector_t injector = info->trap->data;
+    injector->syscall_addr = vdso + syscall_offset;
+
+    return vdso + syscall_offset;
+}
+
+bool setup_post_syscall_trap(drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t syscall_addr)
+{
+    injector_t injector = info->trap->data;
+
+    injector->bp = g_try_malloc0(sizeof(drakvuf_trap_t));
+
+    // setup int3 trap
+    injector->bp->type = BREAKPOINT;
+    injector->bp->name = "injector_post_syscall_trap";
+    // cb will be set from previous call only
+    // we don't have injector_int3_userspace_cb function
+    // in scope here so we will use it from the previous trap
+    injector->bp->cb = info->trap->cb; //injector_int3_userspace_cb;
+    injector->bp->data = injector;
+    injector->bp->breakpoint.lookup_type = LOOKUP_DTB;
+    injector->bp->breakpoint.dtb = info->regs->cr3;
+    injector->bp->breakpoint.addr_type = ADDR_VA;
+    injector->bp->breakpoint.addr = syscall_addr + 2;
+    injector->bp->ttl = UNLIMITED_TTL;
+    injector->bp->ah_cb = NULL;
+
+    if ( drakvuf_add_trap(drakvuf, injector->bp) )
+    {
+        PRINT_DEBUG("Post syscall trap success\n");
+        return true;
+    }
+    else
+    {
+        fprintf(stderr, "Couldn't trap next instruction after syscall\n");
+        return false;
+    }
+
+}
+
+bool save_rip_for_ret(drakvuf_t drakvuf, x86_registers_t* regs)
+{
+
+    // lock vmi
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    ACCESS_CONTEXT(ctx,
+        .translate_mechanism = VMI_TM_PROCESS_DTB,
+        .dtb = regs->cr3
+    );
+    addr_t addr = regs->rsp;
+
+    // make space for storing rip
+    addr -= 0x8;
+    ctx.addr = addr;
+
+    if (VMI_FAILURE == vmi_write_64(vmi, &ctx, &regs->rip))
+    {
+        // release before returning
+        drakvuf_release_vmi(drakvuf);
+        return false;
+    }
+
+    regs->rsp = addr;
```

Comment:
```suggestion
    regs->rsp -= 0x8;
    ACCESS_CONTEXT(ctx,
        .translate_mechanism = VMI_TM_PROCESS_DTB,
        .dtb = regs->cr3,
        .addr = regs->rsp
    );
    if (VMI_FAILURE == vmi_write_64(vmi, &ctx, &regs->rip))
    {
        // release before returning
        drakvuf_release_vmi(drakvuf);
        return false;
    }
```

Reply:
in this, if `vmi_write` fails, then also the `rsp` will change

Reply:
```
bool save_rip_for_ret(drakvuf_t drakvuf, x86_registers_t* regs)
{
    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);

    ACCESS_CONTEXT(ctx,
        .translate_mechanism = VMI_TM_PROCESS_DTB,
        .dtb = regs->cr3,
        .addr = regs->rsp - 0x8
    );

    bool success = false;
    if (VMI_SUCCESS == vmi_write_64(vmi, &ctx, &regs->rip))
    {
        success = true;
        regs->rsp -= 0x8;
    }
    else
        PRINT_DEBUG("Couldn't save rip for ret\n");

    drakvuf_release_vmi(drakvuf);
    return success;
}

```
will this be fine?

Reply:
Yes